Feature/homework 4

seregaby2 opened the pull request:
Home work
You have already started the implementation of your own Web API
application. Based on newly obtained knowledge regarding SOLID
and OOP principles, clean architecture and N Layer architecture
overview and best practices for enterprise solution organization, you
need to reorganize your current implementation to ensure that your
solution is as consistent as possible with the principles and best
practices you learned today.
Also, as this homework should be done after DB creation, you should
update your API to correspond with entities, and their relations
(routes from API perspective).

Commit:
organize solution, add Author mappings

Commit:
reorganize structure folders

Commit:
update book's mapping, add customer, order, orderItem mappings

Commit:
add mapping for DTO's models, moving business logic from controller to services

Commit:
fix some bugs

Commit:
add ServiceCollectionExtensions

Commit:
add enum for book's Genre

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/AuthorService.cs:
```diff
@@ -0,0 +1,74 @@
+            if (existingAuthor == null)
+                return null;
+
+            existingAuthor.FirstName = author.FirstName;
```
It would be nice to add a check for an existing entity.

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/AuthorService.cs:
```diff
@@ -0,0 +1,74 @@
+            author.Id = Guid.NewGuid();
+            author.Books = new List<Book>();
+
+            Authors.Add(author);
```
It would be nice to add a check for an existing entity.

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/AuthorService.cs:
```diff
@@ -0,0 +1,74 @@
+            if (authorToDelete == null)
+                return null;
+
+            Authors.Remove(authorToDelete);
```
When deleting an author, you also need to delete all entities associated with them. For example, all books

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/CustomerService.cs:
```diff
@@ -0,0 +1,76 @@
+            if (existingCustomer == null)
+                return null;
+
+            existingCustomer.Email = customer.Email;
```
Anticipate the error of adding the same customer repeatedly

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/CustomerService.cs:
```diff
@@ -0,0 +1,76 @@
+        {
+            customer.Id = Guid.NewGuid();
+
+            Customers.Add(customer);
```
Anticipate the error of adding the same customer repeatedly

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/CustomerService.cs:
```diff
@@ -0,0 +1,76 @@
+            if (customerToDelete == null)
+                return null;
+
+            Customers.Remove(customerToDelete);
```
When deleting a customer, provide for the deletion of all entities associated with him.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/OrderItemController.cs:
```diff
@@ -37,14 +28,11 @@ public class OrderItemController : ControllerBase
         [ProducesResponseType(typeof(IEnumerable<OrderItemViewDto>), 200)]
         public ActionResult<IEnumerable<OrderItem>> GetAll()
         {
-            var orderItemsDto = _orderItems.Select(a => new OrderItemViewDto
```
As far as I understand, this is a link table between orders and customers. In fact, this controller is not needed, because when creating an order, this entity should be created too. I do not understand the scenario for using this controller.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/AuthorController.cs:
```diff
@@ -61,20 +47,12 @@ public ActionResult<IEnumerable<AuthorViewDto>> GetAll()
         public ActionResult<AuthorViewDto> GetById([FromRoute] Guid id)
         {
-            var author = _authors.FirstOrDefault(a => a.Id == id);
+            var author = _authorService.GetById(id);
```
It would be better to transfer all the logic to services. For example, receiving something. Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/AuthorController.cs:
```diff
@@ -61,20 +47,12 @@ public ActionResult<IEnumerable<AuthorViewDto>> GetAll()
-                BirthDate = author.BirthDate
-            };
 
+            var authorDto = _mapper.Map<AuthorViewDto>(author);
```
It would be better to transfer all the logic to services. For example, receiving something. Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/AuthorController.cs:
```diff
@@ -100,28 +78,13 @@ public ActionResult<AuthorViewDto> GetById([FromRoute] Guid id)
+
+            var author = _mapper.Map<Author>(authorDto);
+
+            var createdAuthor = _authorService.Create(author);
```
It would be better to transfer all the logic to services. For example, receiving something. Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/AuthorController.cs:
```diff
@@ -139,25 +102,17 @@ public ActionResult<AuthorViewDto> Create([FromBody] CreateAuthorDto authorDto)
+            var updatedAuthor = _authorService.Update(id, authorToUpdate);
 
-            return Ok(updatedDto);
+            var authorViewDto = _mapper.Map<AuthorViewDto>(updatedAuthor);
```
Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/BookController.cs:
```diff
@@ -0,0 +1,147 @@
+            if (author == null)
+                return NotFound($"Author with ID {bookDto.AuthorId} was not found.");
+                
+            var book = _mapper.Map<Book>(bookDto);
```
Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/BookController.cs:
```diff
@@ -0,0 +1,147 @@
+
+            var createdBook = _bookService.Create(book);
+
+            var createdDto = _mapper.Map<BookViewDto>(createdBook);
```
Rename please on bookViewDto

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/BookController.cs:
```diff
@@ -0,0 +1,147 @@
+            if (existingBook == null)
+                return NotFound();
+
+            var author = _authorService.GetById(bookDto.AuthorId);
```
Controllers are needed to accept a request, validate it, and send the correct response. The logic for forming the desired model should be in services.

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/BookController.cs:
```diff
@@ -0,0 +1,147 @@
+        [ProducesResponseType(404)]
+        public ActionResult Delete([FromRoute] Guid id)
+        {
+            var bookToDelete = _bookService.Delete(id);
```
Why return a book that was deleted? It is better to return the result of execution, for example true or false

Commit:
delete Model OrderItem(disscused it with mentor)
move validatio to service
delete author with books
return when deleted try or false
check for entity existence on creation (Author, Customer)
add .editorconfig, Apply coding conventions on solution (additional task)

kiryl-kavaleu-lever commented on BookStorage.WebApi/Controllers/CustomerController.cs:
```diff
@@ -0,0 +1,137 @@
+		[ProducesResponseType(404)]
+		public ActionResult<CustomerViewDto> Update([FromRoute] Guid id, [FromBody] UpdateCustomerDto customerDto)
+		{
+			var customerToUpdate = _mapper.Map<Customer>(customerDto);
```
Unnecessary conversion

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/AuthorService.cs:
```diff
@@ -0,0 +1,80 @@
+			bool exists = Authors.Any(a =>
+				string.Equals(a.FirstName, author.FirstName, StringComparison.OrdinalIgnoreCase) &&
+				string.Equals(a.LastName, author.LastName, StringComparison.OrdinalIgnoreCase));
+			if (exists)
```
Please 
Add a space

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/AuthorService.cs:
```diff
@@ -0,0 +1,80 @@
+				return null;
+
+			author.Id = Guid.NewGuid();
+			author.Books = new List<Book>();
```
This entry is not needed

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/BookAuthorService.cs:
```diff
@@ -0,0 +1,26 @@
+		{
+			bool result = _authorService.Delete(authorId);
+
+			if (result)
```
This method will not work. You need to delete a book by BookId. Not by AuthorId. And one author can have several books.

kiryl-kavaleu-lever commented on BookStorage.Application/Implementations/Services/BookService.cs:
```diff
@@ -0,0 +1,95 @@
+
+		public Book? Create(Book book)
+		{
+			var author = _authorService.GetById(book.AuthorID);
```
This method will not work. AuthorId does not come from UI.

kiryl-kavaleu-lever commented on BookStorage.WebApi/DTOs/Customer/CreateCustomerDto.cs:
```diff
@@ -1,4 +1,4 @@
+﻿namespace BookStorage.WebApi.DTOs.Customer
 {
     public class CreateCustomerDto
     {
```
The email field must be unique, required and checked to ensure it is indeed email. Because in the service you have a search for this field.

Commit:
fix some bugs
